src/GafferScene/CollectScenes.cpp

```diff
@@ -41,12 +41,136 @@
 #include "Gaffer/Context.h"
 #include "Gaffer/StringPlug.h"
 
+#include "IECore/NullObject.h"
+
+#include "boost/container/flat_map.hpp"
+
 using namespace std;
 using namespace Imath;
 using namespace IECore;
 using namespace Gaffer;
 using namespace GafferScene;
 
+//////////////////////////////////////////////////////////////////////////
+// RootTree
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+// Used to represent the tree of locations specified by `rootNamesPlug()`.
+// This allows us to quickly find the matching root for any output location
+// via `locationOrAncestor()`.
+class RootTree : public IECore::Data
+{
+
+	public :
+
+		struct Location;
+		using LocationPtr = std::unique_ptr<Location>;
+		using ChildMap = boost::container::flat_map<IECore::InternedString, LocationPtr>;
+
+		struct Location
+		{
+
+			Location( size_t depth )
+				:	depth( depth ), childNames( new InternedStringVectorData() )
+			{
+			}
+
+			bool isRoot() const
+			{
+				return !rootVariableValue.empty();
+			}
+
+			// The path to this location, but exactly as the user spelled
+			// it in `rootNamesPlug()` (may or may not have a leading or
+			// trailing '/'). Empty if not a collection root.
+			string rootVariableValue;
+			size_t depth = 0;
```

Comment:
Well spotted - that dates back to an earlier iteration where I didn't have a constructor for Location. Removed in d2920b9. I've taken the liberty of squashing in rather than pushing a fixup, since it seemed pretty minor.